### PR TITLE
Refactor to use amphp and event-store v8

### PR DESCRIPTION
**Pull request by prolic**
[WIP] knitted with a hot needle - some changes / bugfixes may be required before merge

- This makes prooph/micro even smaller (only 2 files with less than 200 LOC combined).
- 100% async using amphp
- compatible with event-store v8 (async)
- Support async aggregate handlers

How to use this?
- extend a class from `CommandSpecification`, see f.e. `UserSpecification` in the examples dir
- add a command map and create dispatcher, see f.e. `register_and_change_username.php` in the examples dir

Todos:
- [x] Write Tests
- [ ] Support Snapshots (see https://github.com/prooph/micro/issues/53)
- [x] ~~Support Event Publishing~~
- [ ] Documentation (see https://github.com/prooph/micro/issues/52)

ping @codeliner @enumag

As the kernel returns an immutable list of all emitted events, you can simply publish them yourself however you like, that's event a dedicated event publisher will not be part of this. Also in most cases you would rely on event store projections to be running anyway.

### CODE REVIEW

File: examples/Infrastructure/InMemoryEmailGuard.php [hunk 1]
```diff
@@ -20,7 +21,7 @@ final class InMemoryEmailGuard implements UniqueEmailGuard
 
     public function isUnique(string $email): bool
```
Comment (enumag) [inline examples/Infrastructure/InMemoryEmailGuard.php]: Shouldn't this return `Promise<bool>` now?

File: examples/Model/User.php [hunk 2]
```diff
@@ -12,58 +13,52 @@
...
+function registerUser(callable $stateResolver, RegisterUser $command, UniqueEmailGuard $guard): Promise
 {
     if ($guard->isUnique($command->email())) {
-        return [new UserWasRegistered($command->payload())];
+        return new Success(ImmList(new UserRegistered($command->payload())));
     }
 
-    return [new UserWasRegisteredWithDuplicateEmail($command->payload())];
+    return new Success(ImmList(new UserRegisteredWithDuplicateEmail($command->payload())));
 }
```
Comment (enumag) [inline examples/Model/User.php]: I like that these functions are now async and as such can use async guards. But I'm slightly doubtful about it resolving to an `ImmList` of event. In my project I let them return an [asynchronous Iterator](https://amphp.org/amp/iterators/) instead. What is your thought about that?

Note: Code examples assume that UniqueEmailGuard returns a Promise. Also I highly prefer to use the convention of handling error cases first with early return, leaving the happy scenario as main branch so I negated the condition.

```php
function registerUser(callable $stateResolver, RegisterUser $command, UniqueEmailGuard $guard): Iterator
{
    return new Producer(function (callable $emit) use ($command, $guard): Generator {
        if (! yield $guard->isUnique($command->email())) {
             yield $emit(new UserRegisteredWithDuplicateEmail($command->payload()));
             return;
        }
        
        yield $emit(new UserRegistered($command->payload()));
    });
}
```

Admittedly while pure code should return Amp\Iterator here, I later decided that I don't like the `Producer + $emit` syntax so I actually let these functions return `Generator<Promise|EventInterface>`. Then there is a wrapper function that transforms it into `\Amp\Iterator<EventInterface>`. I consider such syntactic improvement as acceptable over code purity for my project although not sure if it's suitable for an example like this... But I'll show it to you anyway for comparison:

```php
function registerUser(callable $stateResolver, RegisterUser $command, UniqueEmailGuard $guard): Generator
{
    if (! yield $guard->isUnique($command->email())) {
         yield new UserRegisteredWithDuplicateEmail($command->payload());
         return;
    }

    yield new UserRegistered($command->payload());
}
```

The wrapper function looks something like this:

```php
function transform(Generator $generator): Iterator
{
    return new Producer(function (callable $emit) use ($generator) {
        foreach ($generator as $eventOrPromise) {
            if ($eventOrPromise instanceof Promise) {
                yield $eventOrPromise;
            } else {
                yield $emit($eventOrPromise);
            }
        }
    });
}
```

File: examples/Infrastructure/InMemoryEmailGuard.php (hunk 1 above)
Comment (prolic) [reply examples/Infrastructure/InMemoryEmailGuard.php]: yes

File: examples/Model/User.php (hunk 2 above)
Comment (prolic) [reply examples/Model/User.php]: awesome, thanks !!

**Comment by prolic** [comment]
@enumag changes done

**Comment by codeliner** [comment]
@prolic the implementation looks really great. Hopefully I find some time to try it out. I'd also look at a way to connect [Event Engine Cockpit](https://github.com/event-engine/cockpit) to it.

File: examples/Model/User.php (hunk 2 above)
Comment (prolic) [reply examples/Model/User.php]: @enumag I ran into issues with this. On `yield $eventOrPromise;` the promises is yielded, but not assigned to the variable in the handler function.
`$state = yield $stateResolver();` results in `$state = null`.

Comment (enumag) [reply examples/Model/User.php]: Ah yeah... sorry, I copy pasted a wrong function. I'll post the correct one tomorrow.

Comment (prolic) [reply examples/Model/User.php]: thanks

Comment (prolic) [reply examples/Model/User.php]: maybe can you directly submit a PR to this branch?

File: src/Kernel.php [hunk 3]
```diff
@@ -44,44 +46,66 @@ function buildCommandDispatcher(
...
             try {
-                $es = yield $specification->handle(stateResolver($eventStore, $specification));
+                $iterator = new Producer(function (callable $emit) use ($eventStore, $specification, $readBatchSize): Generator {
+                    foreach ($specification->handle(stateResolver($eventStore, $specification, $readBatchSize)) as $eventOrPromise) {
+                        if ($eventOrPromise instanceof Promise) {
+                            yield $eventOrPromise;
+                        } else {
+                            yield $emit($eventOrPromise);
+                        }
+                    }
```
Comment (enumag) [inline src/Kernel.php]:
```suggestion
                    $generator = $specification->handle(stateResolver($eventStore, $specification, $readBatchSize));
                    while ($generator->valid()) {
                        $eventOrPromise = $generator->current();

                        if ($eventOrPromise instanceof Promise) {
                            $generator->send(yield $eventOrPromise);
                        } else {
                            yield $emit($eventOrPromise);

                            $generator->next();
                        }
                    }
```

Comment (enumag) [reply src/Kernel.php]: @prolic This should work I think. ;-)

Comment (prolic) [reply src/Kernel.php]: Awesome, thank you.

**Comment by prolic** [comment]
I added new issues for snapshot support and documentation. I'll merge this to master.

**Pull request merged**